### ISSUE TITLE
[Issue]: Asymmetric HIP peer access causes P2P/SHM transport disagreement

### ISSUE THREAD

**Issue by hyjuunn** [ROCm/rocm-systems#10586]
### Problem Description

RCCL can select different transports at the two endpoints of one connection when `hipDeviceCanAccessPeer()` reports support in only one direction. One endpoint selects SHM while the other selects P2P/IPC. Communicator setup then fails because the opaque connection record produced by one transport is decoded as the other transport's structure.

Versions referenced below:

| Purpose | Version |
|---|---|
| Initial observation | RCCL 2.30.4 (installed) |
| Two-GPU control/patched A/B | Public source build, commit `704600e2b8` |
| Current candidate revision | RCCL 2.30.7 headers and library |

Two 8x Radeon Pro W7900 (`gfx1100`) systems of the same ASRock Rack model are referenced below:

- **System A** — used for the BIOS before/after platform study.
- **System B** — used for the RCCL control/patched A/B.

Both exhibited asymmetric peer capability. They differed in how many BAR apertures the firmware placed within reach.

## Observed behavior

System B exposed seven one-way GPU pairs, 21 no-access pairs, and no mutual-access pairs. One representative pair reported:

```text
hipDeviceCanAccessPeer(0, 1) = 0
hipDeviceCanAccessPeer(1, 0) = 1
```

The two ranks selected incompatible transports for that pair:

```text
rank 0: 0 -> 1 via SHM/direct/direct
rank 1: 1 -> 0 via P2P/IPC
```

The P2P endpoint then failed with:

```text
[FATAL ERROR]: HIP failure: 'invalid argument'
```

The SHM endpoint interpreted the P2P connection bytes as SHM metadata and reported an impossible segment size:

```text
Error while attaching to shared memory segment /dev/shm/nccl- (size 91628832292864), error: No such file or directory (2)
```

The varying multi-terabyte size is not a real allocation request. `ncclConnect` is an opaque byte record that each transport casts to its own connection type. Once the endpoints choose different transports, P2P connection bytes are read as an SHM descriptor on the other endpoint.

No-access pairs work because both endpoints reject P2P and fall back together. The failure is specific to a one-way pair, where the endpoints disagree.

## Root cause in RCCL

`p2pCanConnect()` performs one peer-capability query:

```cpp
// hipDeviceCanAccessPeer() on AMD
cudaDeviceCanAccessPeer(&p2p, cudaDev1, cudaDev2);
```

Each endpoint calls transport selection with its own device as the first argument, so the two ends of one connection evaluate the same pair in opposite orders. With a one-way pair, both answers are individually valid but produce incompatible transport choices.

In the observed failure, the directional peer-capability query caused the endpoints to reach different transport decisions. The INFO traces show the same connection being constructed as SHM on one endpoint and P2P/IPC on the other. Requiring mutual peer capability removes that disagreement.

## Why the platform reported asymmetric capability

Directionality is part of the `hipDeviceCanAccessPeer()` API contract; an asymmetric result is not, by itself, a HIP API violation.

On both systems, the production BIOS used an MMIO Above-4G ceiling of `0x3ffbffffffff` (approximately 64 TiB). Each W7900 exposed a 44-bit DMA mask, with a maximum address of `0x0fffffffffff` (below 16 TiB). The firmware could therefore place 64 GiB GPU BAR0 apertures above the address range reachable by another W7900. The BIOS exposed no configurable MMIO-limit option.

ASRock supplied a test BIOS that changed the Above-4G limit to `0x0fffffffffff`. A before/after test on System A showed:

| Check | Original BIOS 10.04 | Test BIOS |
|---|---:|---:|
| W7900 devices detected | 8 | 8 |
| 64 GiB BAR0 apertures below 16 TiB | 2/8 | 8/8 |
| Mutual unordered HIP pairs | 1/28 | 28/28 |
| One-way unordered HIP pairs | 12/28 | 0/28 |
| No-access unordered HIP pairs | 15/28 | 0/28 |

After the firmware update, raw HIP kernels on System A read 1,048,576 floats successfully in all 56 directed off-diagonal paths with zero mismatches, and an exact-value eight-rank AllGather passed with P2P enabled.

On System B, only one BAR was below 16 TiB, which produced the seven one-way / 21 no-access / zero mutual topology described above. Removing `pci=realloc=on` in a controlled reboot did not change any BAR base or peer-matrix entry, so that kernel flag was not the cause. The firmware MMIO policy was the platform trigger.

This agrees with the ROCm BAR documentation:

> If the memory address of a BAR memory exceeds the physical addressing limit
> of a device, the device will not be able to access that BAR.

https://rocm.docs.amd.com/en/docs-6.4.1/how-to/Bar-Memory.html

The same document recommends placing the high-MMIO aperture within the device's physical-address limit by configuring `MMIO High Base`, `MMIO High size`, and Above 4G Decoding. The production BIOS exposed no configurable MMIO-limit option, so that solution could not be applied through its setup interface.

Correcting the firmware removes this specific trigger, but RCCL should still make one consistent transport decision whenever a platform legitimately reports directional peer capability.

## Expected behavior

Both endpoints must select a compatible transport. A pair should use P2P only when the capability required by both endpoints is available; otherwise both endpoints should consistently fall back.

### Operating System

Ubuntu 24.04 LTS (Noble Numbat)

### CPU

2 x AMD EPYC 9554 64-Core Processor

### GPU

8 x AMD Radeon Pro W7900 Dual Slot (gfx1100)

### ROCm Version

ROCm 7.14.0 (HIP runtime 7.14.60850)

### ROCm Component

rccl

### Steps to Reproduce

1. Identify a one-way pair. Logical GPU ordering can change across boots, so use the matrix from the current boot:

```python
import torch

n = torch.cuda.device_count()
for a in range(n):
    for b in range(a + 1, n):
        fwd = torch.cuda.can_device_access_peer(a, b)
        rev = torch.cuda.can_device_access_peer(b, a)
        if fwd != rev:
            print(f"one-way pair: {a}->{b}={int(fwd)} {b}->{a}={int(rev)}")
```

2. Run an exact-value two-rank AllGather with `NCCL_P2P_DISABLE` unset:

```bash
env -u NCCL_P2P_DISABLE \
  HIP_VISIBLE_DEVICES=<gpu-a>,<gpu-b> \
  NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,P2P \
  torchrun --standalone --nproc_per_node=2 ag_correctness.py
```
Contents of `ag_correctness.py`:

```python
import torch
import torch.distributed as dist

dist.init_process_group("nccl")
rank = dist.get_rank()
world = dist.get_world_size()
torch.cuda.set_device(rank)

elements = 1024 * 1024
source = torch.full(
    (elements,), rank, dtype=torch.bfloat16, device=f"cuda:{rank}"
)
gathered = torch.empty(
    world * elements, dtype=source.dtype, device=source.device
)

dist.all_gather_into_tensor(gathered, source)
torch.cuda.synchronize()

if rank == 0:
    chunks = gathered.view(world, elements).float()
    means = chunks.mean(dim=1).cpu().tolist()
    minima = chunks.amin(dim=1).cpu().tolist()
    maxima = chunks.amax(dim=1).cpu().tolist()
    expected = torch.arange(world, dtype=torch.float32)

    print(f"means={means}")
    print(f"minima={minima}")
    print(f"maxima={maxima}")
    print(f"expected={expected.tolist()}")
    print(f"correct={torch.equal(torch.tensor(means), expected)}")

dist.destroy_process_group()
```
3. Observe one rank selecting SHM, the other selecting P2P/IPC, followed by communicator initialization failure.
4. As a control, a pair with no access in either direction falls back to SHM on both endpoints and completes.

## Candidate fix and result

In the AMD branch of `p2pCanConnect()`, query the reverse direction when the forward query reports support, and retain P2P only if both directions report support. This adds no collective data-path work; it runs only during transport initialization.

A two-GPU control/patched A/B from the same source commit (`704600e2b8`) and a separate eight-GPU run of the equivalent patch produced:

| Test | Source revision | Result |
|---|---|---|
| Unmodified, one-way 2-GPU pair | `704600e2b8` | Mixed SHM/P2P transports; HIP invalid argument; failed |
| Patched, same 2-GPU pair | `704600e2b8` | SHM selected at both endpoints; collective completed |
| Patched, all 8 GPUs | `96a25b5f` | All logged ring connections used SHM, 0 P2P connections; exact values `[0..7]`; `correct=True` |

After the firmware correction made all pairs mutual, installed RCCL retained direct P2P (`isAllDirectP2p=1`) and public `rccl-tests` completed with `#wrong=0` and zero out-of-bounds values.

The current candidate revision (`b3c9ffeb81`) was also tested with `rccl-tests` rebuilt against its matching 2.30.7 headers and library. The exact candidate library retained `P2P/direct pointer`, passed the eight-GPU in-place and
out-of-place AllGather sweep with zero wrong and zero out-of-bounds values, and exited successfully.

Candidate implementation: #10554 

## Workaround

`NCCL_P2P_DISABLE=1` globally avoids the mismatched P2P decision, at the cost of disabling P2P for every pair. The candidate fix is narrower: it rejects only pairs that do not report mutual peer capability.

## Supporting evidence

I have raw logs for the firmware before/after comparison, the unpatched/patched RCCL A/B, the eight-rank validation, and the exact current candidate revision. I can provide them if they would be useful for further investigation.
